### PR TITLE
Add narrow workaround for FreeBSD clang aliasing

### PR DESCRIPTION
**Pull request by svaarala**
Rough solution to make FreeBSD + -m32 + clang prior to 5.0 to avoid clang aliasing issues automatically. This avoids unnecessary segfaults due to broken union assignments etc. Though they are detectable with -DDUK_USE_SELF_TESTS, it takes some diagnosis to figure them out.

### CODE REVIEW

**Comment by michael-o** [comment]
Besides the missing `__`, tried this from master, ran `dist.py`, `configure.py` and `Makefile.cmdline`:

    [mosipov@mika-ion ~/Projekte/duktape/dist]$ make -f Makefile.cmdline
    clang -o duk  -Os -pedantic -std=c99 -Wall -fstrict-aliasing -fomit-frame-pointer -I./examples/cmdline -I./src -DDUK_CMDLINE_PRINTALERT_SUPPORT -I./extras/print-alert -DDUK_CMDLINE_CONSOLE_SUPPORT -I./extras/console -DDUK_CMDLINE_LOGGING_SUPPORT -I./extras/logging -DDUK_CMDLINE_MODULE_SUPPORT -I./extras/module-duktape src/duktape.c examples/cmdline/duk_cmdline.c extras/print-alert/duk_print_alert.c extras/console/duk_console.c extras/logging/duk_logging.c extras/module-duktape/duk_module_duktape.c -lm
    [mosipov@mika-ion ~/Projekte/duktape/dist]$ ./duk
    Segmentation fault (Speicherabzug geschrieben)

Where is the error in my approach?

File: config/header-snippets/packed_tval_fillin.h.in [hunk 1]
```diff
@@ -31,10 +31,18 @@
 #endif
 #endif
 
+/* FreeBSD, -m32, and clang prior to 5.0 has union aliasing issues which
+ * break duk_tval copying.  Disable packed duk_tval automatically.
+ */
+#if defined(DUK_F_FREEBSD) && defined(DUK_F_X86) \
+    defined(__clang__) && defined(__clang_major__) && (__clang_major < 5)
```
Comment (michael-o) [inline config/header-snippets/packed_tval_fillin.h.in]: `__clang_major` => ` __clang_major__`

Comment (svaarala) [reply config/header-snippets/packed_tval_fillin.h.in]: Thanks, fixed!

**Comment by svaarala** [comment]
@michael-o Apart from the __clang_major__ typo, did this work out in your case?

**Comment by michael-o** [comment]
@svaarala Unfortunately not, as I have mentioned [here](https://github.com/svaarala/duktape/pull/1764#issuecomment-331701488). Did I perform the wrong steps?

**Comment by svaarala** [comment]
@michael-o Doh, the diff was missing a `&&`, fixed and pushed now (I don't have FreeBSD so I'm doing this blindly... which is hard :).

**Comment by michael-o** [comment]
I already fixed the `&&` locally, but this still does not work. This is nested in another `#ifdef`: `#if !defined(DUK_F_PACKED_TVAL_PROVIDED)`. I think that this is defined because I have added `#error` and it was not invoked. Hint: I ran `configure.py` with default options. Question: this should run out of the box. Why is this define present?

**Comment by svaarala** [comment]
DUK_F_PACKED_TVAL_PROVIDED is necessary to get proper "tri-state" behavior for DUK_USE_PACKED_TVAL:

DUK_USE_PACKED_TVAL may be defined or undefined by architecture, compiler, or platform files. If they do that, configure.py will automatically emit DUK_F_PACKED_TVAL_PROVIDED so that it won't get re-defined or re-undefined. The main problem is that without such a define, it's not possible to differentiate between "DUK_USE_PACKED_TVAL not defined" vs. "DUK_USE_PACKED_TVAL explicitly #undef'd" for the purposes of defaulting.

In this pull the current problem is that the architecture_x86.h.in file defines:

```
#define DUK_USE_ARCH_STRING "x86"
#if !defined(DUK_USE_BYTEORDER)
#define DUK_USE_BYTEORDER 1
#endif
/* XXX: This is technically not guaranteed because it's possible to configure
 * an x86 to require aligned accesses with Alignment Check (AC) flag.
 */
#if !defined(DUK_USE_ALIGN_BY)
#define DUK_USE_ALIGN_BY 1
#endif
#define DUK_USE_PACKED_TVAL
```

This causes DUK_F_PACKED_TVAL_PROVIDED to be defined and the default check to be bypassed. I'll try moving the workaround to the x86 architecture file.

**Comment by svaarala** [comment]
Ok, moved the workaround to the x86 architecture file. It's not an ideal place if this affects any other 32-bit architectures. Unfortunately feature detection is never really very clean because sometimes the reasons to enable/disable something may come from the architecture, the compiler, the platform, etc, so there's often no single right place for something.

**Comment by michael-o** [comment]
Testing...

**Comment by michael-o** [comment]
Hooray:

    [mosipov@mika-ion ~/Projekte/duktape/dist]$ make -f Makefile.cmdline CC=cc
    cc -o duk  -Os -pedantic -std=c99 -Wall -fstrict-aliasing -fomit-frame-pointer -I./examples/cmdline -I./src -DDUK_CMDLINE_PRINTALERT_SUPPORT -I./extras/print-alert -DDUK_CMDLINE_CONSOLE_SUPPORT -I./extras/console -DDUK_CMDLINE_LOGGING_SUPPORT -I./extras/logging -DDUK_CMDLINE_MODULE_SUPPORT -I./extras/module-duktape src/duktape.c examples/cmdline/duk_cmdline.c extras/print-alert/duk_print_alert.c extras/console/duk_console.c extras/logging/duk_logging.c extras/module-duktape/duk_module_duktape.c -lm
    [mosipov@mika-ion ~/Projekte/duktape/dist]$ ./duk
    ((o) Duktape 2.2.99 (v2.2.0-14-g3fc72d47)
    duk> new Date();
    = "2017-09-24T19:49:42.813Z"
    duk>

**Comment by svaarala** [comment]
If you have time, it'd be nice if you could compile with clang 4.x and 5.x (with configure.py defaults) and check the output for:

```
$ duk -e 'print(Duktape.env)'
```

The env string indicates (in a pretty cryptic fashion) whether duk_tval is packed (8 bytes) or unpacked (16 bytes). In this case it should be unpacked for clang 4.x and packed for clang 5.x.

**Comment by michael-o** [comment]
Sure thing, will do for 4.0.0, 4.0.1 and 5.0.0 for you.

**Comment by michael-o** [comment]
[mosipov@mika-ion ~/Projekte/duktape/dist]$ make -f Makefile.cmdline CC=cc
    cc -o duk  -Os -pedantic -std=c99 -Wall -fstrict-aliasing -fomit-frame-pointer -I./examples/cmdline -I./src -DDUK_CMDLINE_PRINTALERT_SUPPORT -I./extras/print-alert -DDUK_CMDLINE_CONSOLE_SUPPORT -I./extras/console -DDUK_CMDLINE_LOGGING_SUPPORT -I./extras/logging -DDUK_CMDLINE_MODULE_SUPPORT -I./extras/module-duktape src/duktape.c examples/cmdline/duk_cmdline.c extras/print-alert/duk_print_alert.c extras/console/duk_console.c extras/logging/duk_logging.c extras/module-duktape/duk_module_duktape.c -lm
    [mosipov@mika-ion ~/Projekte/duktape/dist]$ ./duk -e 'print(Duktape.env)'
    ll u n p1 a1 x86 bsd clang
    [mosipov@mika-ion ~/Projekte/duktape/dist]$ rm ./duk && make -f Makefile.cmdline CC=clang40
    clang40 -o duk  -Os -pedantic -std=c99 -Wall -fstrict-aliasing -fomit-frame-pointer -I./examples/cmdline -I./src -DDUK_CMDLINE_PRINTALERT_SUPPORT -I./extras/print-alert -DDUK_CMDLINE_CONSOLE_SUPPORT -I./extras/console -DDUK_CMDLINE_LOGGING_SUPPORT -I./extras/logging -DDUK_CMDLINE_MODULE_SUPPORT -I./extras/module-duktape src/duktape.c examples/cmdline/duk_cmdline.c extras/print-alert/duk_print_alert.c extras/console/duk_console.c extras/logging/duk_logging.c extras/module-duktape/duk_module_duktape.c -lm
    [mosipov@mika-ion ~/Projekte/duktape/dist]$ ./duk -e 'print(Duktape.env)'
    ll u n p1 a1 x86 bsd clang
    [mosipov@mika-ion ~/Projekte/duktape/dist]$ rm ./duk && make -f Makefile.cmdline CC=clang50
    clang50 -o duk  -Os -pedantic -std=c99 -Wall -fstrict-aliasing -fomit-frame-pointer -I./examples/cmdline -I./src -DDUK_CMDLINE_PRINTALERT_SUPPORT -I./extras/print-alert -DDUK_CMDLINE_CONSOLE_SUPPORT -I./extras/console -DDUK_CMDLINE_LOGGING_SUPPORT -I./extras/logging -DDUK_CMDLINE_MODULE_SUPPORT -I./extras/module-duktape src/duktape.c examples/cmdline/duk_cmdline.c extras/print-alert/duk_print_alert.c extras/console/duk_console.c extras/logging/duk_logging.c extras/module-duktape/duk_module_duktape.c -lm
    [mosipov@mika-ion ~/Projekte/duktape/dist]$ ./duk -e 'print(Duktape.env)'
    ll p n p1 a1 x86 bsd clang

**Comment by svaarala** [comment]
Looks good, duk_tval is packed for clang 5.0 and unpacked before it.

**Comment by michael-o** [comment]
I guess this one is good to merge now?

**Comment by svaarala** [comment]
There are various other options for dealing with this too. If a compiler option could fix the issue and allow packed duk_tval to be used prior to Clang 5.0, one might prefer e.g. adding a specific self test for FreeBSD and/or Clang, and cause a descriptive fatal error explaining the possible workarounds. User could then choose between (a) compiler option workaround and getting packed duk_tval, (b) Duktape option workaround and getting (more memory consuming) unpacked duk_tval, or (c) upgrading their Clang.

OTOH, if no compiler option workaround is possible, then the user options are really just -UDUK_USE_PACKED_TVAL (which this workaround does automatically) or upgrading their compiler.

All in all, while quite ugly, this merge would make sense to me because it'd save unnecessary segfault / selftest diagnosis for FreeBSD.

**Comment by michael-o** [comment]
I am looking at this from a FreeBSD ports perspective. I don't know wether the fix from 5 will ever be backported, but what we can do is to modify the port in a way that it will check for presence of Clang 5 or provide the options to compile with 5. Though, I don't want to impose the user to install another compiler for a single feature to save some memory.

**Comment by svaarala** [comment]
From that ports perspective, if you can do a compiler version check and can detect x86 vs. x64, then you can just use `configure.py [...] -UDUK_USE_PACKED_TVAL` when the x86 + clang <5.0 condition is met, even without the workaround in this pull.

With this pull that check would no longer be necessary.

**Comment by michael-o** [comment]
That is correct, but it would require to depend on Python at build time. I'd like to avoid additional complexity in the port's Makefile.

**Comment by svaarala** [comment]
Ok, I see, you're using the default sources to avoid a Python 2 dependency and this workaround would make this easier.

Unless it's a big hassle, I'd still recommend allowing configure.py to run, because one may want to tweak something in the configuration at some point. For instance, `-DDUK_USE_FASTINT` improves integer performance but it is not enabled by default because it makes some portability assumptions that are quite OK for mainline platforms but can be problematic for embedded (esp. bare metal) or very old targets.

**Comment by michael-o** [comment]
Right, such options can be added easily to the port, though it makes it more complex to maintain. As long as people won't ask, I won't add those. I'd be happy to have the header files, C source, shared lib and the executable in `/usr/local` to continue with the Gerbera port for starters. I will get back to you as soon as the port has been updated by my patch.

**Comment by svaarala** [comment]
I can fully understand that. Good luck with the rest of the port!

**Comment by michael-o** [comment]
I did another test in i386 with 10.4-PRERELEASE and Clang 3.4.1:

    $ uname -a
    FreeBSD <hostname> 10.4-PRERELEASE FreeBSD 10.4-PRERELEASE #8 r323607: Fri Sep 15 10:38:24 CEST 2017     root@<hostname>:/usr/obj/usr/src/sys/BLNN719X  i386
    duktape/dist$ cc --version
    FreeBSD clang version 3.4.1 (tags/RELEASE_34/dot1-final 208032) 20140512
    Target: i386-unknown-freebsd10.4
    Thread model: posix
    $ make -f Makefile.cmdline CC=cc
    cc -o duk  -Os -pedantic -std=c99 -Wall -fstrict-aliasing -fomit-frame-pointer -I./examples/cmdline -I./src -DDUK_CMDLINE_PRINTALERT_SUPPORT -I./extras/print-alert -DDUK_CMDLINE_CONSOLE_SUPPORT -I./extras/console -DDUK_CMDLINE_LOGGING_SUPPORT -I./extras/logging -DDUK_CMDLINE_MODULE_SUPPORT -I./extras/module-duktape src/duktape.c examples/cmdline/duk_cmdline.c extras/print-alert/duk_print_alert.c extras/console/duk_console.c extras/logging/duk_logging.c extras/module-duktape/duk_module_duktape.c -lm
    duktape/dist$ ./duk -e 'print(Duktape.env)'
    ll u n p1 a1 x86 bsd clang

**Pull request merged**